CMake/FindFlatBuffers.cmake

```diff
@@ -16,24 +16,49 @@
 # Find the flatbuffers schema compiler
 #
 # Output Variables:
-# * FLATBUFFERS_FLATC_EXECUTABLE the flatc compiler executable
-# * FLATBUFFERS_FOUND
+# * FLATBUFFERS_FLATC_EXECUTABLE - The flatc compiler executable
+# * FLATBUFFERS_FLATC_VERSION - The version of flatc found.
+# * FlatBuffers_FOUND - Whether flatc was found.
+# * FLATBUFFERS_FOUND - Legacy alias of the above (all caps).
 #
-# Provides:
-# * FLATBUFFERS_GENERATE_C_HEADERS(Name <files>) creates the C++ headers
-#   for the given flatbuffer schema files.
-#   Returns the header files in ${Name}_OUTPUTS
+# If flatc is found, the following imported target is created:
+# * flatbuffers::flatc - Imported target for the compiler
 
 set(FLATBUFFERS_CMAKE_DIR ${CMAKE_CURRENT_LIST_DIR})
 
 find_program(FLATBUFFERS_FLATC_EXECUTABLE NAMES flatc)
-find_path(FLATBUFFERS_INCLUDE_DIR NAMES flatbuffers/flatbuffers.h)
+
+if(EXISTS ${FLATBUFFERS_FLATC_EXECUTABLE})
+  # detect version
+  execute_process(COMMAND ${FLATBUFFERS_FLATC_EXECUTABLE} --version
+    RESULT_VARIABLE FLATBUFFERS_FLATC_VERSION_RESULT
+    OUTPUT_VARIABLE FLATBUFFERS_FLATC_VERSION_OUTPUT)
+
+  if(FLATBUFFERS_FLATC_VERSION_RESULT EQUAL 0)
+    # The output looks like "flatc version 23.3.3", so use a regex to trim out the part we need
+    string(REGEX REPLACE "flatc version ([0-9]+\\.[0-9]+\\.[0-9]+).*" "\\1" FLATBUFFERS_FLATC_VERSION ${FLATBUFFERS_FLATC_VERSION_OUTPUT})
+  else()
+    message(WARNING "Failed to execute flatc to check version")
+  endif()
+endif()
 
 include(FindPackageHandleStandardArgs)
 find_package_handle_standard_args(FlatBuffers
-  DEFAULT_MSG FLATBUFFERS_FLATC_EXECUTABLE FLATBUFFERS_INCLUDE_DIR)
+  VERSION_VAR FLATBUFFERS_FLATC_VERSION
+  REQUIRED_VARS FLATBUFFERS_FLATC_EXECUTABLE)
 
-if(FLATBUFFERS_FOUND)
+# CMake standard requires the found var to match the case of the filename.
+# Provide legacy alias for all-caps.
+set(FLATBUFFERS_FOUND ${FlatBuffers_FOUND})
+
+if(FlatBuffers_FOUND)
+  # Provide imported target for the executable
+  add_executable(flatbuffers::flatc IMPORTED GLOBAL)
```

Comment:
FindFlatBuffers now provides an imported target `flatbuffers::flatc`, in line with CMake best practices